### ISSUE TITLE
YoloV2 doesn't actually work correctly after change to latest darknet with YoloV3 release?

### ISSUE THREAD

**Issue by sudeepmandal** [AlexeyAB/darknet#1810]
Firstly, a million thanks to @AlexeyAB for maintaining this repo and bringing some sanity to the world of Darknet/Yolo.

I have been using YoloV2 for a while on a custom problem, since before YoloV3 was released with associated updates to darknet. Ever since Darknet was updated with the release of YoloV3, the YoloV2 models never work as well any more. There seems to be a serious bug as I have consistently seen a large number of nonsensical false alarms that never occurred with the previous darknet + YoloV2. I can take the same training data, cfg file and original pre-trained weights (darknet19_448.conv.23) and get great performance with an older fork of this repo, whereas with the latest version, even though training appears to go well with reducing training loss, the test performance is much poorer than with the older fork (2a9f7e44ce1b73d3d56ef83f83e94f074ecac3f9)

Hard for me to provide a reproducible test case, but it appears that hardly anyone seems to be using YoloV2 after V3 and associated darknet changes were released. 

Are we supposed to adapt our old YoloV2 annotations/config files, etc, in some way due to changes in darknet with the release of YoloV3? I've tried retraining my YoloV2 models with latest version of this repo every few months and performance is always much poorer than with the performance with the same setup on a fork from the commit mentioned above.

**Comment by AlexeyAB** [comment]
@sudeepmandal Hi,

* What accuracy (mAP) can you get for old and new repo?

* Can you get good accuracy for default weights files http://pjreddie.com/media/files/yolo-voc.weights and https://github.com/AlexeyAB/darknet/blob/master/cfg/yolov2-voc.cfg ?

* If you train your own network with old version, then can you get good accuracy for detection on the new version?

* What GPU, CUDA and cuDNN versions do you use?

**Comment by sudeepmandal** [comment]
Those are good questions @AlexeyAB . I will spend some time investigating and try to get back to you soon.

For your last question, I am using:
GTX 1080 Ti, CUDA 9.1, cuDNN 9.1 and openCV 3.4.0

**Comment by sudeepmandal** [comment]
@AlexeyAB - I think that the issue I am seeing is that when I run a trained model on a video, the generated video frame and bbox annotation is out of sync by what appears to be one frame. In fact, the annotated bboxes are one frame ahead of the actual video frame in the output video which results in many instances where the bboxes are ahead of the actual classes that are moving in the frame. Visually, it appears as though the trained model is not performing well since there are many frames that you visually notice as having class bboxes with no object within (since the video frame is lagging the annotation display).

I typically train my final models with all my annotated frames for a custom application since I don't have thousands of annotated frames, and then just run it on a video and eye-ball the model performance on the video (which consists mostly of frames the model has not seen). With the newer darknet, it appears that this bug may have resulted in me assuming that the trained model was at fault, but it might simply be a bug in the annotation and writing out of an annotated video when running `./darknet detector demo`.

**Comment by AlexeyAB** [comment]
> In fact, the annotated bboxes are one frame ahead of the actual video frame in the output video

It was in old version too. To disable it, change this line: https://github.com/AlexeyAB/darknet/blob/31ac46ba22112c8122d84a6a2b00db3a1ecb92cb/src/demo.c#L18
to this
`#define FRAMES 1`

Try to check accuracy (mAP) on Training or/and Test datasets on Old and New Darknet.

**Comment by sudeepmandal** [comment]
That's interesting, I checked and my old fork also has `#define FRAMES 3` but didn't have this issue in the generated videos. I ran the mAP calculations on the train images with the new model (using the current latest version of your repo) and the numbers look good ( > 95% mAP ) for training data. 

Is there a way to generated annotated images for a sequence of test images or do I need to just write a for-loop in bash to do it? That way I can test to see if old and new models are generating similar outputs on unseen frames.

Thanks!

**Comment by AlexeyAB** [comment]
> Is there a way to generated annotated images for a sequence of test images or do I need to just write a for-loop in bash to do it? That way I can test to see if old and new models are generating similar outputs on unseen frames.

What do you mean?

You can collect some images (frames from video) by yourself, then run this command on Old and New repo and compare `results.txt` files:
`darknet.exe detector test data/voc.data yolo-voc.cfg yolo-voc.weights -dont_show -ext_output < data/train.txt > result.txt`

Or just compare mAP for Training images. If it will the same, then for unseen frames it will be the same.

**Comment by sudeepmandal** [comment]
Ah okay. Thanks. I will try that. FYI, I think making the change you mentioned above did solve the issue with the latest repo in terms of video being in sync. Is there a reason you have still left it at 3 for now?

I'll experiment with the models and their outputs tomorrow. I appreciate all your help in troubleshooting this!

**Comment by AlexeyAB** [comment]
>  Is there a reason you have still left it at 3 for now?

In the realtime videos the frames isn't much different from each other. So it improve detection quality on the most of videos.

**Comment by sudeepmandal** [comment]
> In the realtime videos the frames isn't much different from each other. So it improve detection quality on the most of videos.

Ah okay. That makes sense. In my case, each frame is taken from a 1 FPS stream, so the latency makes a big visual impact in terms of accuracy. Thanks a lot for all your assistance and for the great work on this repo! 👍 

I'll close this issue now since I think this addresses the main issue I was having. The new and old models seem to be performing similarly at this point. I will continue to investigate and re-open in case anything else comes up related to performance with new darknet vs. my old fork. Thanks!
